Still update list if filter fails with MatcherException

dennisschagt opened the pull request:
Resolves https://github.com/newsboat/newsboat/issues/2857

This is one way to resolve the issue. It shows the resulting list after failing to apply a filter (empty because it directly fails at the first feed).
An alternative option would be to rollback the change in filter (possibly applying the previous filter if one was active). I didn't check, but I expect that would be a bit more error prone (which I rather don't do just before a release).
Yet another way to fix would be to avoid this case of MatcherException (non-existing attribute) altogether (by doing a trial run for one item or by adding machinery to check the available attributes while parsing the filter)

Minoru commented:
Oh, I'd love to check the expressions before they run! That's definitely *not* for the upcoming release though :)

This reminds me of https://github.com/newsboat/newsboat/issues/335 and https://github.com/newsboat/newsboat/issues/1053. Heaps of work though, and #1053 is blocked on https://github.com/newsboat/newsboat/issues/631 which is blocked on https://github.com/newsboat/newsboat/issues/1344 which is blocked on https://github.com/newsboat/newsboat/issues/2326 (a lot of content passes through filters, so we must guarantee there won't be UTF-8-related crashes).

I think doing the trial run is the best option. The message "applying the filter failed" implies (to me at least) that the state of the program didn't change, so trial run is an expected behaviour. I wonder what we'd do if the list is empty though.

As for approach in this PR, I think it's an okay stopgap as it puts Newsboat into a more useful state, but I'm disinclined to merge it because the bug doesn't seem too severe to warrant a stopgap. I think it's okay to live with it for one more release, and add a proper solution in the meantime.

dennisschagt commented:
Rebased onto master to get build results for FreeBSD 14.1.

> I think doing the trial run is the best option. The message "applying the filter failed" implies (to me at least) that the state of the program didn't change, so trial run is an expected behaviour.

Implemented a trial run, resulting message:
![image](https://github.com/user-attachments/assets/9a946870-e182-4c74-a11a-7b6f87029c7e)

Might also need to be implemented for the article list.
However, non-existing attributes are less likely to occur there (all feed attributes are also usable in article list filters, which is not the case the other way around)

The current solution is not perfect at stopping issues because the trial run might not hit a non-existing attribute if the filter expression short-circuits on the test feed. (e.g. `define-filter "short-circuit test"  "latest_article_age < 100 or nonexistingattribute = 5"`)

> Oh, I'd love to check the expressions before they run!

Nice, I do have some ideas so might try one of these ideas in the next few days.
